internal_packages/keybase/lib/encrypt-button.cjsx

```diff
@@ -0,0 +1,140 @@
+pgp = require 'kbpgp'
+_ = require 'underscore'
+
+class EncryptMessageButton extends React.Component
+
+  @displayName: 'EncryptMessageButton'
+
+  # require that we have a draft object available
+  @propTypes:
+    draftClientId: React.PropTypes.string.isRequired
+
+  constructor: (props) ->
+    super(props)
+    @state = @_getStateFromStores()
+
+    # maintain the state of the button's toggle
+    @state.currentlyEncrypted = false
+
+    # store the message's plaintext in case the user wants to edit further after
+    # hitting the "encrypt" button (i.e. so we can "undo" the encryption)
+    @state.plaintext = ""
+
+    # store the message's body here, for comparison purposes (so that if the
+    # user edits an encrypted message, we can revert it)
+    @state.cryptotext = ""
+
+  componentDidMount: ->
+    @unlistenKeystore = PGPKeyStore.listen(@_onKeystoreChange, @)
+    # periodically check to see if we should get any keys
+    @interval = setInterval(@_onTrigger, 100)
+
+  componentWillUnmount: ->
+    @unlistenKeystore()
+    clearInterval(@interval)
+
+  _getStateFromStores: ->
+    keys = []
+    DraftStore.sessionForClientId(@props.draftClientId).then (session) =>
+      for recipient in session.draft().to
+        publicKeys = PGPKeyStore.pubKeys(recipient.email)
+        if publicKeys.length < 1
+          # no key for this user:
+          # push a null so that @_encrypt can line this array up with the
+          # array of recipients
+          keys.push({address: recipient.email, key: null})
+        else
+          # note: this, by default, encrypts using every public key associated
+          # with the address
+          for publicKey in publicKeys
+            if not publicKey.key?
+              PGPKeyStore.getKeyContents(key: publicKey)
+            else
+              keys.push(publicKey)
+
+    return {
+      keys: keys
+    }
+
+  _onKeystoreChange: =>
+    if @state.currentlyEncrypted
+      # message is no longer encrypted with all keys - decrypt it!
+      @_toggleCrypt()
+    # now actually get the new keys
+    @setState(@_getStateFromStores())
+
+  _onTrigger: =>
+    # Some sort of outside event has occured that we must respond to
+    DraftStore.sessionForClientId(@props.draftClientId).then (session) =>
+      if @state.currentlyEncrypted
+        # THERE CAN BE NO CHANGES
+        if @state.cryptotext? and @state.cryptotext != ""
+          # can't put HTML in or it will be sent with the message, confusing recipients
+          #body = '<div style="background-color: #D3D3D3; padding: 10px;">' + @state.cryptotext + '</div>'
+          body = @state.cryptotext
+          session.changes.add({body: body})
+      else
+        plaintext = session.draft().body
+        @setState({plaintext: plaintext})
+
+    @setState(@_getStateFromStores())
+
+  render: ->
+    <div className="n1-keybase">
+      <button title="Encrypt email body" className="btn btn-toolbar" onClick={ => @_onClick()} ref="button">
```

Comment:
We should be able to make this `onClick={@_onClick}`. Making an extra little bound function that calls onClick shouldn't be necessary.